authenticator/server/config/config.go

```diff
@@ -27,194 +125,232 @@ type Config struct {
 	LogRaw    bool
 
 	SecretsManager string
-	ConfigFilePath string
 
-	// Special flags
-	devMode bool
-	Version bool
-
-	// Vault specific flags
+	// Vault related
 	VaultToken     string
 	VaultTokenPath string
 	VaultAddr      string
 
-	// AWS Secrets Manager specific flags
+	// AWS secrets manager related
 	AwsRegion string
+
+	// Special flags
+	ConfigFilePath string
+	DevMode        bool
+	Version        bool
+}
+
+func verify(config *Config) error {
+	if !config.DisableTLS {
+		if config.PathToTLSCert == "" {
+			return errors.New("tls is enabled but no tls cert has been provided")
+		}
+		if config.PathToTLSKey == "" {
+			return errors.New("tls is enabled but no tls key has been provided")
+		}
+	}
+	return nil
 }
 
 // Parse returns the parsed config. A pointer is not returned
 // because after first parse, the config is immutable.
 func Parse() (Config, error) {
-	config, err := parse()
+	config, err := parseWithPrecedence()
 	if err != nil {
 		return Config{}, err
 	}
-	if err := verify(config); err != nil {
-		return Config{}, err
-	}
-	if config.devMode {
-		// dev mode uses the file back-end
-		if err := os.Unsetenv("VAULT_ADDR"); err != nil {
+	if config.DevMode {
+		config, err = devModeConfig()
+		if err != nil {
 			return Config{}, err
 		}
-		config = Config{
-			Host:       "127.0.0.1",
-			HTTPPort:   6000,
-			GRPCPort:   6001,
-			DisableTLS: true,
-			LogLevel:   "debug",
-			LogFormat:  "text",
-			LogRaw:     false,
-		}
-	}
-	return config, nil
-}
-
-func parse() (Config, error) {
-	var config Config
-	setConfigDefaults()
-	if err := setConfigFlags(); err != nil {
-		return Config{}, err
 	}
-	if err := setConfigEnvVars(); err != nil {
+	if err := verify(config); err != nil {
 		return Config{}, err
 	}
-	if err := viper.Unmarshal(&config); err != nil {
-		return Config{}, err
+	// We return a non-pointer to signal that it will never change.
+	return *config, nil
+}
+
+func parseWithPrecedence() (*Config, error) {
+	flag.Parse()
+
+	// Our order of precedence is:
+	// 1. Env var
```

Comment:
Why should env var take precedence over command line flag? Would this lead to unintended behavior since flags are more explicitly provided in comparison to env vars?

Reply:
Hm, that's a good point. I think that logic is very sound. I will switch them! :-)

Reply:
https://github.com/cyralinc/approzium/pull/180/commits/b28649f5d70cfa988a25df24b21e78f5bfca3e7f